### DISCUSSION TITLE
A new way to handle language changes

### DISCUSSION THREAD

**Discussion by jtheisen** [dotnet/csharplang#6945] [Language Ideas] [answered]
This is more a meta-suggestion about how new language features could be experimented with, but it still requires changes to Roslyn.

I have a strong C++ background and I'm a bit worried that C# follows it's fate into ever-more complexity. C++ is at a point where, if I had to run a team and needed C or C++, I'd prefer C just to avoid endless discussions about "what kind of C++" were to be coded and about it's countless quirks and traps. It's not that I don't like the new C# features, but some of them are quite opinionated and I think C# is in a spot that is particularly well suited to try a different approach to creating new language features.

Inspired by the code generators feature that now allows code generation to be rolled out as NuGet packages, I wonder if the same couldn't be done with new language features.

I'm not familiar with Roslyn's sources, but maybe there's a good way to provide extensions points to allow changes to the language be rolled out as NuGet packages as well. That way, future features could be offered without requiring consumers to change the compiler - it would just be a matter of installing the extensions from NuGet, in the same way it's currently done for code generators and analyzers.

If that's technically possible, the advantage would be that new features could stay longer in the experimental phase where projects that like to use bleeding-edge features can use them "in the wild" without the need to push them to the official language immediately. During that phase, a lot more experience could be collected with these features to decide whether they really are the way to go.

If the features don't make it, the projects that already adopted them can keep using them as long as the extension will be maintained.

On the other hand, it also encourages experimentation by a greater number of people, because it's easier to find users that would be willing to try a new feature if it's just a matter of installing a package.

Just food for thought.

**Comment by HaloFour** [comment]
C# already supports having features in "preview" which requires being explicitly enabled within a project, that does enable a feature to be put out for the public to use without official support and to gather feedback.

**Comment by TahirAhmadov** [comment, marked as answer]
I think the problem would be maintaining essentially a large number of versions, depending on the mix and match of various Nuget packages. Language changes often introduce ramifications to other parts of the language, and it's not as trivial as adding or removing a referenced library, for example. It's plenty of work just maintaining effectively 2 versions - production and preview - as it is.

**Comment by erikhermansson79** [comment]
I like the new way of using preview features. I would like to have a version window where obsolete features outside the window are treated as errors. That way the language team can guide users to a more modern usage of the language but still keep backwards compatability for those who need it.

**Reply by HaloFour** [reply]
> I would like to have a version window where obsolete features outside the window are treated as errors. That way the language team can guide users to a more modern usage of the language but still keep backwards compatability for those who need it.

That would imply that the language team considers any existing features to be obsolete.
